cliquet/storage/memory.py

```diff
@@ -180,11 +184,28 @@ def _bump_timestamp(self, collection_id, parent_id):
             the time will slide into the future. It is not problematic since
             the timestamp notion is opaque, and behaves like a revision number.
         """
+        is_specified = (record is not None
+                        and modified_field in record
+                        or last_modified is not None)
+        if is_specified:
+            # If there is a timestamp in the new record, try to use it.
+            if last_modified is not None:
+                current = last_modified
+            else:
+                current = record[modified_field]
+        else:
+            current = utils.msec_time()
+
         previous = self._timestamps[collection_id].get(parent_id)
-        current = utils.msec_time()
         if previous and previous >= current:
-            current = previous + 1
-        self._timestamps[collection_id][parent_id] = current
+            collection_timestamp = previous + 1
+        else:
+            collection_timestamp = current
+
+        if not is_specified:
+            current = collection_timestamp
+
+        self._timestamps[collection_id][parent_id] = collection_timestamp
```

Comment:
nit: Since the flow of this function is a bit complex, I would suggest putting a summary of the spec in the docstring (or a bit more comments).
It has been only a few days since we paired on this, and I already feel unsure about the intentions :D